### PR TITLE
Allow cron::Schedules as function arguments

### PR DESCRIPTION
**Pull request by hgzimmerman**
This PR depends on [this change](https://github.com/zslayton/cron/issues/98) in the `cron` crate making its way to crates.io and should not be merged before then. The minimum version is anticipated to be either 0.9.1 or 0.10.0.

This PR should allow taking schedule arguments as either `&str` or `Schedule` types. This is preferable over taking plain `&str`s because it allows testing that the cron strings are valid before passing them to your library, lowering the chance of a programmer error causing a scheduled task to not run when ignoring errors returned from constructing `Job`s

### CODE REVIEW

**Comment by mvniekerk** [comment]
Hi @hgzimmerman Thank you for the PR. I see it is still marked as WIP?

**Comment by hgzimmerman** [comment]
Yup, I just renamed the PR to indicate it is a WIP. It may take a bit for the change in cron to be released, given the last release was a year ago, so this may just sit here for a while.

File: src/job.rs [hunk 1]
```diff
@@ -382,12 +383,13 @@ impl JobLocked {
     /// sched.add(job)
     /// tokio::spawn(sched.start());
     /// ```
-    pub fn new<T>(schedule: &str, run: T) -> Result<Self, Box<dyn std::error::Error>>
+    pub fn new<S, T>(schedule: S, run: T) -> Result<Self, Box<dyn std::error::Error>>
     where
         T: 'static,
         T: FnMut(Uuid, JobsSchedulerLocked) + Send + Sync,
+        S: TryInto<Schedule, Error=cron::error::Error>
```
Comment (hgzimmerman) [inline src/job.rs:390]: `Error=cron::error::Error` won't work here. The function needs to take an `E: std::error::Error` type parameter in order to handle both the `str` parameter, as well as the blanket impl for `TryFrom<U: Into<T>> for T`, which has `std::convert::Infallible` as the associated `Error` type. (https://doc.rust-lang.org/std/convert/trait.TryFrom.html#impl-TryFrom%3CU%3E)

**Comment by mvniekerk** [comment]
Hey @hgzimmerman 
Kindly check - I've merged a huge code change into main for 0.6-beta.1.
There's an error enum JobSchedulerError that also may be of interest for you.

**Comment by hgzimmerman** [comment]
I'm finding the places where support for taking a `TryInto<Schedule, E>` would be added return an error that can only come from failing to parse the `&str` to a `Schedule`. 
Instead, if duplicate functions that only took Schedule were added, they wouldn't need to return a `Result`.
So you would have something like:

```rust
    pub fn new_infallible<T>(schedule: Schedule, run: T) -> Self
    where
        T: 'static,
        T: FnMut(Uuid, JobsSchedulerLocked) + Send + Sync,
    {
        let schedule: Schedule = schedule.try_into()?;
        let job_id = Uuid::new_v4();
        Self(Arc::new(RwLock::new(Box::new(CronJob {
            data: JobStoredData {
               // ...
            },
            run: Box::new(run),
            run_async: Box::new(nop_async),
            async_job: false,
        }))))
    }
    pub fn new<T>(schedule: &str, run: T) -> Result<Self, Box<dyn std::error::Error>>
    where
        T: 'static,
        T: FnMut(Uuid, JobsSchedulerLocked) + Send + Sync,
    {
        let schedule = Schedule::from_str(schedule)?
        Ok(Self::new_infallible(schedule, run))
    }
```

This could be done without waiting for `cron` to release to crates.io, but would require adding functions with somewhat odd naming schemes that do the same thing as the current ones, but return `Self` instead of `Result<Self, Box<dyn Error>>`. 

Do you have any thoughts regarding this versus what is currently in this PR?

**Comment by BlackDex** [comment]
@hgzimmerman looks like cron is updated :tada: https://github.com/zslayton/cron/pull/99

https://github.com/zslayton/cron/releases/tag/v0.10.0

**Comment by mvniekerk** [comment]
Hey @hgzimmerman & @BlackDex 
Looking forward getting this patch in. We'll chat after 0.6 beta 2 is released? I'm writing example code for the Nats integration then I'll be releasing it.

**Comment by mvniekerk** [comment]
Okay so 0.6 beta 2 is out.
You'll see that job.rs moved to job/mod.rs
If you can kindly update your PR?

**Comment by mvniekerk** [comment]
Thank you @hgzimmerman  for your contribution. I'll merge the PR and have it part of the 0.6 release 🙌🏻

**Pull request merged**